cpp/src/arrow/dataset/file_base.h

```diff
@@ -45,82 +46,150 @@ namespace dataset {
 /// be read like a file
 class ARROW_DS_EXPORT FileSource {
  public:
-  // NOTE(kszucs): it'd be better to separate the BufferSource from FileSource
-  enum SourceKind { PATH, BUFFER };
-
   FileSource(std::string path, std::shared_ptr<fs::FileSystem> filesystem,
-             Compression::type compression = Compression::UNCOMPRESSED,
-             bool writable = true)
+             Compression::type compression = Compression::UNCOMPRESSED)
```

Comment:
I'm not satisfied how this class is transforming into a Franken-class. The need of static fake properties and the semi-broken default constructor.

I'd say make FileSource an interface and use inheritance, make Open() virtual, the `path()` and `filesystem()`  will be specific to one implementation (maybe name them Source, FileSource, BufferSource, ...). We can make an accept visitor for classes who wants to touch properties like the path.

Reply:
I agree that `FileSource` would benefit from refactoring but I think that doing so further in this PR (including resorbing `WritableFileSource`) is not necessary. I'll make a follow up JIRA